### PR TITLE
Cndeep

### PR DESCRIPTION
**Pull request by acpaquette**
Progress so far on deepening points in autocnet control networks. Currently, the method is able to find points to deepen based on previously computed matches that have been thrown out but are still valid.

To do: 
- Clean up current methods and funcs. Try to limit how much nodes and edges are passed around
- Find some metric that can accurately find the best potential match out of the keypoints that exist that have not been matched together. That is, for points who do not have a match in the matches dataframe, how do we find the best potential punch through point? More importantly, does that point even exist on the other image?

Other to dos can be found in the methods that need work.

[Link to Notebook with this functionality](https://gist.github.com/acpaquette/a1f2cf5a2b6fc7172a6ff530831f9db6)

### CODE REVIEW

File: autocnet/control/control.py [hunk 1]
```diff
@@ -90,8 +91,125 @@ def overlaps(group):
...
+        # Same as above but all calculations are source, destination dependent
+        keypoint = edge.destination.get_keypoint_coordinates(index = keypoint_idx, homogeneous=True).values
+        correspondences = func(keypoint, keypoints, edge, source, destination, *args, **kwargs)
+        kp_id_matches = matches.loc[matches['destination_idx'] == keypoint_idx]
+        edge_id_matches = kp_id_matches[kp_id_matches['source_idx'].isin(correspondences)]
+        dist_matches = edge_id_matches[edge_id_matches['distance'] < dist]
+        if any(dist_matches):
+            return matches.loc[dist_matches.loc[:, ['distance']].idxmin()].source_idx.values
+        else:
+            subpixel_keypoints = edge.source.get_keypoint_coordinates(index = correspondences).values
```
Comment (Kelvinrr) [inline autocnet/control/control.py]: @jlaura I think we need a definite solution to this source/destination problem?

File: autocnet/utils/utils.py [hunk 2]
```diff
@@ -455,3 +455,37 @@ def wrapper(*args, **kwarg):
...
+def dest_to_source(kp, source, destination, homogeneous = False):
+    """
+    Using a kp, an edge source, and an edge destination. Translate that keypoints
+    from destination to source.
+    """
+    destin_latlon = destination.geodata.pixel_to_latlon(kp[0], kp[1])
+    if homogeneous:
+        return source.geodata.latlon_to_pixel(destin_latlon[0], destin_latlon[1]) + (1,)
+    else:
+        return source.geodata.latlon_to_pixel(destin_latlon[0], destin_latlon[1])
```
Comment (Kelvinrr) [inline autocnet/utils/utils.py]: Probably a good idea to factor out `source.geodata.latlon_to_pixel(destin_latlon[0], destin_latlon[1])`

File: autocnet/matcher/deepen_funcs.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    Returns
+    ----------
+    : list
+      A list of points that lie within the geometric threshold of the epipolar
+      line.
+    """
+    if source < destination:
+        f_matrix = edge['fundamental_matrix'].T
+    else:
+        f_matrix = edge['fundamental_matrix']
```
Comment (Kelvinrr) [inline autocnet/matcher/deepen_funcs.py]: put a comment mentioning why the transposing is necessary

File: autocnet/graph/edge.py [hunk 4]
```diff
@@ -523,5 +523,5 @@ def get_matches(self): # pragma: no cover
         dkps.columns = ['destination_x', 'destination_y']
         match = match.join(skps, on='source_idx')
         match = match.join(dkps, on='destination_idx')
-        matches.append(match)
-        return matches
+        # matches.append(match)
```
Comment (Kelvinrr) [inline autocnet/graph/edge.py]: remove?

File: autocnet/control/control.py [hunk 5]
```diff
@@ -90,8 +91,125 @@ def overlaps(group):
...
+        keypoint = edge.source.get_keypoint_coordinates(index = keypoint_idx, homogeneous=True).values
+        # Use some function to limit the number of correspondences
+        correspondences = func(keypoint, keypoints, edge, source, destination, *args, **kwargs)
+        # Check if those correspondences exist as a match within some distance threshold
+        kp_id_matches = matches.loc[matches['source_idx'] == keypoint_idx]
+        edge_id_matches = kp_id_matches[kp_id_matches['destination_idx'].isin(correspondences)]
+        dist_matches = edge_id_matches[edge_id_matches['distance'] < dist]
+        # If there is some reasonable correspondence return the found match index
+        if any(dist_matches):
+            return matches.loc[dist_matches.loc[:, ['distance']].idxmin()].destination_idx.values
```
Comment (Kelvinrr) [inline autocnet/control/control.py]: maybe put it in more than one line to make it slightly more readable.

**Comment by Kelvinrr** [comment]
Looks like Tutorials With Vis is being removed, I imagine we want to keep it in.

**Comment by jlaura** [comment]
@acpaquette The ultimate decision is that this is most likely going to need to use tri-focal tensors?

**Comment by jlaura** [comment]
Let's wait on this for now.  We have (maybe I do in a notebook and I need to push it) epipolar punch through working (mathematically), but still with noise. I want to use the trifocal tensor to be doing this work, but suspect open a new PR specific to the TT would be best?

**Comment by acpaquette** [comment]
Agreed, I just wanted to get this updated and the notebook I was working in linked.

**Pull request closed without merging**